---
title: "atacManual"
author: "Jordan S. Kesner"
date: "December 8, 2018"
output: html_document
---

Run the analysis components of the pipeline one at a time manually


Run a single footprint
```{r}

bampath <- "/home/rstudio2/atac/h508/bam/h508_all_dedup.bam"
baipath <- "/home/rstudio2/atac/h508/bam/h508_all_dedup.bam.bai"
motif_score <- "90%"
upstream <- 100
downstream <- 100
scope <- paste0("chr", c(1:22, "X", "Y"))
genome <- Hsapiens


symbol <- "TCF7"

organism_rows = grep('Hsapiens', values(MotifDb)$organism, ignore.case = TRUE)
gene_symbol_rows = grep(symbol, values(MotifDb)$geneSymbol, ignore.case = TRUE)
human_gene_rows = intersect(gene_symbol_rows, organism_rows)
gene_motifs <- as.list(MotifDb[human_gene_rows])
unique_gene_motifs <- unique(gene_motifs)

PWM <- unique_gene_motifs[[1]]
  
savefile <- "/home/rstudio2/atac/h508/graphs/tcf7_motif1.png"
png(filename = savefile)
    
    # generate signal
    signal <- factorFootprints(
                            # bam file input
                            bamfiles = bampath,
                            # bai index input
                            index = baipath,
                            # PWM input
                            pfm = PWM,
                            # reference genome
                            genome = genome,
                            # minimum motif matching score
                            min.score = motif_score,
                            # scope of analysis
                            seqlev = scope,
                            # bp upstream of motif to analyze
                            upstream = upstream,
                            # bp downstream of motif to analyze
                            downstream = downstream
                            )
    # turn off, important
    dev.off()


```


Make a single heatmap (ChipPeakAnno) library
```{r}

featureAlignedHeatmap(signal$signal, 
                      feature.gr=reCenterPeaks(signal$bindingSites,
                                               width=200+width(signal$bindingSites[1])), 
                      annoMcols="score",
                      sortBy="score",
                      n.tile=ncol(signal$signal[[1]]))

```


