---
title: "atacFParacne"
author: "Jordan S. Kesner"
date: "January 29, 2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

https://web.mit.edu/~r/current/arch/i386_linux26/lib/R/library/S4Vectors/html/Rle-class.html

```{r}
#install.packages("BiocManager")
#BiocManager::install("aracne.networks", version = "3.8")
#BiocManager::install("mygene", version = "3.8")
#BiocManager::install("chromoMap", version = "3.8")
#
library(ATACseqQC)
library(GenomicRanges)
library(aracne.networks)
library(mygene)
library(chromoMap)
```

Load the required data
```{r}

#
load("C:\\Users\\jsk33\\Desktop\\LS1034-WT-01.CBFA2T2.motif1.Rdata")
l <- ls()
l <- l[-8]
rm(list = l)
rm(l)
#
wg_sites <- FPdata[["wg_bindingsites"]]

#
chromosome <- wg_sites@seqnames
sites <- wg_sites@ranges

# convert the Rle factor object to a linear vector
chromosome <- as.vector(chromosome, mode="any")

```

Step 1 - Build a GRanges object based on the predicted ARACNe targets of current gene
```{r}

# get the ARACNe interactome
coad_interactome <- aracne.networks::reguloncoad

# get the mnx1 targets
mnx1_targets <- names(coad_interactome[["9139"]][["tfmode"]])

# Retrieve info for the network targets
target_list <- list()
for (a in 1:length(mnx1_targets)){
  target_list[a] <- mygene::getGene(geneid = mnx1_targets[a], fields = "all")}

# count the number of gene locations we have
loc_count <- 0
for (a in 1:length(mnx1_targets)){
  if (is.null(target_list[[a]][["genomic_pos"]])){next} else {
  if (is.list(target_list[[a]][["genomic_pos"]][[1]])){
    loc_count <- (loc_count + length(target_list[[a]][["genomic_pos"]]))
  } else {
    loc_count <- (loc_count + 1)}}}

# Retrieve the genomic coordinates of all network targets
# retrieve all genomic coords here, can later trim non standard ones more easily with GRanges functions
target_locations <- matrix(data = NA, nrow = loc_count, ncol = 4)
colnames(target_locations) <- c("gene", "chr", "start", "end")
idx <- 1
#
for (a in 1:loc_count){
  
  if (is.null(target_list[[a]][["genomic_pos"]])){next} else {
    
    if (is.list(target_list[[a]][["genomic_pos"]][[1]])){
    
      for (b in 1:length(target_list[[a]][["genomic_pos"]])){
        target_locations[idx,1] <- target_list[[a]][["symbol"]]
        target_locations[idx,2] <- paste0("chr", target_list[[a]][["genomic_pos"]][[b]][["chr"]])
        target_locations[idx,3] <- target_list[[a]][["genomic_pos"]][[b]][["start"]]
        target_locations[idx,4] <- target_list[[a]][["genomic_pos"]][[b]][["end"]]
        idx <- (idx+1)
        } # end for (b in 1:length(target_list[[a]][["genomic_pos"]]))

        } else {
    
        target_locations[idx,1] <- target_list[[a]][["symbol"]]
        target_locations[idx,2] <- paste0("chr", target_list[[a]][["genomic_pos"]][["chr"]])
        target_locations[idx,3] <- target_list[[a]][["genomic_pos"]][["start"]]
        target_locations[idx,4] <- target_list[[a]][["genomic_pos"]][["end"]]
        idx <- (idx+1)}}}

## Convert the genomic locations of the targets into GRanges
gr <- GRanges(
              seqnames = target_locations[,2],
              ranges = IRanges(start = as.numeric(target_locations[,3]), end = as.numeric(target_locations[,4])))
              #strand = c(rep("+", times = num_names)),
              #seqinfo = Seqinfo(genome="hg38"),
              #score = c(rep(1, times = num_names)))

# prune to standard xsomes
gr <- keepStandardChromosomes(gr, pruning.mode="coarse")


## Intersect the targets GRanges with the binding sites list
intersection <- intersect(gr, wg_sites)

```

 

Code testing / sanity checks
```{r}

canno <- data.frame(
  name = c(1:148),
  chrom = as.vector(gr@seqnames, mode="any"),
  start = gr@ranges@start,
  data = gr@ranges@width)


cmap <- chromoMap(canno, type = "annotation")
cmap <- chromoMap(gr, type ="heatmap-single", HeatColRange = c("blue","white","red"))



gr1 <- GRanges(
  seqnames = c("chr1", "chr2"),
  ranges = IRanges(start = c(1000, 1000), end = c(2000, 2000))
)

gr2 <- GRanges(
  seqnames = c("chr1", "chr3"),
  ranges = IRanges(start = c(1900, 1000), end = c(3000, 2000))
)

gr3 <- intersect(gr1, gr2)


```




